ssz/toolz/dump.py

```diff
@@ -0,0 +1,107 @@
+from typing import (
+    Mapping,
+    Sequence,
+)
+
+from eth_utils import (
+    to_dict,
+)
+
+from ssz.sedes import (
+    BaseSedes,
+    Boolean,
+    ByteList,
+    ByteVector,
+    Container,
+    List,
+    Serializable,
+    UInt,
+    Vector,
+)
+from ssz.sedes.serializable import (
+    MetaSerializable,
+)
+
+from .formatters import (
+    DefaultFormatter,
+)
+
+
+def to_formatted_dict(value, sedes=None, basic_sedes_formatter=DefaultFormatter):
+    return dump(value, sedes, basic_sedes_formatter)
+
+
+def dump(value, sedes=None, basic_sedes_formatter=DefaultFormatter):
+    if sedes is None:
+        if isinstance(value, Serializable):
```

Comment:
Te way the current code is structured, if someone called `dump(Thing(...), sedes=Thing)` we would end up hitting the final `else` block and throw a `TypeError`.  There should be a test that demonstrates this.  I think this behavior would be surprising.

I would advocate for restructuring this as:

```python
if sedes is None:
    if isinstance(value, Serializable):
        sedes = type(value)
    else:
        raise ...

if isinstance(sedes, Serializable):
    dump_serializable(value, sedes, ...)
elif ...
```

This structure *allows* for you to try and dump a `Serializable` using some other sedes that isn't its class.  If this isn't desirable, then we can tighten the restriction and say that `isinstance(value, Serializable) and sedes is not type(value)` is an error case, though we should be explicit about raising a properly worded error for it to provide good error messages.

Reply:
The `dump(Thing(...), sedes=Thing)` case is handled in 
```
elif isinstance(sedes, MetaSerializable):
        return dump_serializable(value, basic_sedes_formatter)
```
I'll add a test to demenstrate it.

Reply:
Ah, I missed the `MetaSerializable` part.  :+1:

Reply:
I am still of the opinion that this function would be easier to read if it was done as:

```
def dump():
   if sedes is None:
       # This code only raises errors or sets the `sedes` value.

    ```
    # All of the "dump" code is here, fully separate from the error raising code above.
```

Probably a minor win and entirely possible that I was just reading hastily... do as you please.

Reply:
Ah, missed this one, will fix